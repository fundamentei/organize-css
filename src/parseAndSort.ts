import generate from "@babel/generator";
import * as babel from "@babel/parser";
import traverse from "@babel/traverse";
import {
  ExpressionStatement,
  isArrowFunctionExpression,
  isExpressionStatement,
  isIdentifier,
  isObjectExpression,
  isStringLiteral,
  ObjectMethod,
  ObjectProperty,
  SpreadElement
} from "@babel/types";
import groups from "./groups";
import pseudo from "./groups/pseudo";
import hyphenateStyleName from "./hyphenateStyleName";

type TProperties = Array<ObjectMethod | ObjectProperty | SpreadElement>;

function getStyleNameFromObjectProperty(property: ObjectProperty): string | null {
  if (isIdentifier(property.key)) {
    return hyphenateStyleName(property.key.name);
  }

  if (isStringLiteral(property.key)) {
    return pseudo.indexOf(property.key.value) !== -1 ? property.key.value : hyphenateStyleName(property.key.value);
  }

  return null;
}

function getSortedProperties(properties: TProperties): TProperties {
  const rational: TProperties = [];
  const missing: TProperties = [];

  // those properties will be placed at the end of the object in the order they're declared
  const footer: TProperties = properties.filter(property => {
    if (property.type === "SpreadElement" || property.type === "ObjectMethod") {
      return true;
    }

    if (property.type === "ObjectProperty") {
      return property.computed;
    }

    return false;
  });

  const knownProperties = groups.reduce((a, b) => a.concat(b), []);

  properties.forEach(property => {
    if (property.type !== "ObjectProperty") {
      return;
    }

    const key = getStyleNameFromObjectProperty(property);

    // those properties are the ones we don't know how to sort...so keep them in the `missing` array in the order that
    // they're declared and they'll will be put back in the end of the object
    if (key && knownProperties.indexOf(key) === -1) {
      missing.push(property);
    }
  });

  groups.forEach(cssProperties => {
    cssProperties.forEach(cssProperty => {
      const property = properties.find(property => {
        if (footer.indexOf(property) !== -1) {
          return false;
        }

        if (property.type === "ObjectProperty") {
          const styleName = getStyleNameFromObjectProperty(property);
          return styleName === cssProperty;
        }

        return false;
      });

      if (property) {
        rational.push(property);
      }
    });
  });

  return rational.concat(missing).concat(footer);
}

export default async function parseAndSort(objectExpression: string): Promise<string | null> {
  try {
    const ast = await babel.parse(`() => (${objectExpression})`, {
      sourceType: "module"
    });

    if (!ast) {
      return null;
    }

    traverse(ast, {
      ObjectExpression: path => {
        path.node.properties = getSortedProperties(path.node.properties);
      }
    });

    if (!isExpressionStatement(ast.program.body[0])) {
      return null;
    }

    const expression = (ast.program.body[0] as ExpressionStatement).expression;

    if (!isArrowFunctionExpression(expression) || !isObjectExpression(expression.body)) {
      return null;
    }

    const generated = generate(expression.body);
    return generated ? generated.code : null;
  } catch (err) {
    return null;
  }
}
