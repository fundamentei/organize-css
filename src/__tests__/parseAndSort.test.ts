import * as fs from "fs";
import parseAndSort from "../parseAndSort";

describe("parseAndSort", () => {
  it.each(["1.txt", "2.txt", "3.txt"])("works for the case %s", async filename => {
    const withSortedCSSProperties = await parseAndSort(loadCase(filename));
    expect(withSortedCSSProperties).toMatchSnapshot();
  });

  it("is null if the JS couldn't be parsed", async () => {
    const withSortedCSSProperties = await parseAndSort("[}");
    expect(withSortedCSSProperties).toBe(null);
  });

  it("keeps spread elements in the order that they're declared", async () => {
    const withSortedCSSProperties = await parseAndSort(`{ ...a, display: "flex", ...b, ...c }`);
    expect(withSortedCSSProperties).toMatchSnapshot();
  });

  it("doesn't touch computed properties", async () => {
    const withSortedCSSProperties = await parseAndSort(`{ ...a, [display]: "flex", ...b, ...c }`);
    expect(withSortedCSSProperties).toMatchSnapshot();
  });

  it("sorts nested declarations", async () => {
    const withSortedCSSProperties = await parseAndSort(
      JSON.stringify({
        top: 0,
        position: "absolute",
        left: 0,
        bottom: 0,
        right: 0,
        "&:before": {
          top: 0,
          position: "absolute",
          left: 0,
          bottom: 0,
          right: 0,
          "&:hover": {
            top: 0,
            position: "absolute",
            left: 0,
            bottom: 0,
            right: 0
          }
        }
      })
    );

    expect(eval(`(${withSortedCSSProperties})`)).toStrictEqual({
      position: "absolute",
      top: 0,
      right: 0,
      bottom: 0,
      left: 0,
      "&:before": {
        position: "absolute",
        top: 0,
        right: 0,
        bottom: 0,
        left: 0,
        "&:hover": {
          position: "absolute",
          top: 0,
          right: 0,
          bottom: 0,
          left: 0
        }
      }
    });
  });

  it("doesn't remove properties that it doesn't know how to sort", async () => {
    const withSortedCSSProperties = await parseAndSort(
      JSON.stringify({
        WebkitAppearance: "none",
        display: "flex"
      })
    );

    expect(eval(`(${withSortedCSSProperties})`)).toStrictEqual({
      display: "flex",
      WebkitAppearance: "none"
    });
  });
});

function loadCase(filename: string) {
  return fs.readFileSync(`${__dirname}/cases/${filename}`).toString("utf8");
}
