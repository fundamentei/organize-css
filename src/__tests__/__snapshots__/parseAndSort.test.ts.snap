// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseAndSort doesn't touch computed properties 1`] = `
"{ ...a,
  [display]: \\"flex\\",
  ...b,
  ...c
}"
`;

exports[`parseAndSort keeps spread elements in the order that they're declared 1`] = `
"{
  display: \\"flex\\",
  ...a,
  ...b,
  ...c
}"
`;

exports[`parseAndSort works for the case 1.txt 1`] = `
"{
  position: \\"absolute\\",
  top: 0,
  right: 0,
  bottom: 0,
  left: 0,
  zIndex: 1,
  display: \\"inline-flex\\",
  alignItems: \\"center\\",
  justifyContent: \\"center\\",
  boxSizing: \\"border-box\\",
  minWidth: 18,
  maxWidth: 18,
  minHeight: 18,
  maxHeight: 18,
  padding: 2,
  backgroundColor: disabled ? palette.catskillWhite : palette.white,
  border: \`solid 1px \${palette.mystic}\`,
  borderRadius: 2,
  cursor: \\"pointer\\",
  pointerEvents: disabled ? \\"none\\" : \\"all\\",
  \\"&:hover\\": {
    border: \`solid 1px \${palette.perano}\`
  },
  \\"&:focus\\": {
    border: \`solid 1px \${shades.primary.variants[6]}\`,
    outline: \\"none\\",
    boxShadow: \`\${shades.primary.variants[8]} 0px 0px 0px 3px\`
  }
}"
`;

exports[`parseAndSort works for the case 2.txt 1`] = `
"{
  position: \\"absolute\\",
  top: 0,
  right: 0,
  bottom: 0,
  left: 0,
  zIndex: 1,
  display: \\"inline-flex\\",
  alignItems: \\"center\\",
  justifyContent: \\"center\\",
  boxSizing: \\"border-box\\",
  minWidth: 18,
  maxWidth: 18,
  minHeight: 18,
  maxHeight: 18,
  padding: 2,
  borderRadius: 2,
  cursor: \\"pointer\\"
}"
`;

exports[`parseAndSort works for the case 3.txt 1`] = `
"{
  position: \\"absolute\\",
  right: 0,
  ...defaultStyles,
  ...getCSSStyles(disabled)
}"
`;
